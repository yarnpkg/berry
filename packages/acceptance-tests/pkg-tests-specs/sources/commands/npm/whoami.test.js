const {
  fs: {writeFile},
  tests: {startPackageServer},
} = require(`pkg-tests-core`);

const AUTH_TOKEN = `686159dc-64b3-413e-a244-2de2b8d1c36f`;
const AUTH_TOKEN2 = `316158de-64b3-413e-a244-2de2b8d1c80f`;
const AUTH_IDENT = `dXNlcm5hbWU6YSB2ZXJ5IHNlY3VyZSBwYXNzd29yZA==`;

const INVALID_AUTH_TOKEN = `a24cb960-e6a5-45fc-b9ab-0f9fe0aaae57`;
const INVALID_AUTH_IDENT = `dXNlcm5hbWU6bm90IHRoZSByaWdodCBwYXNzd29yZA==`; // username:not the right password

describe(`Commands`, () => {
  describe(`npm whoami`, () => {
    test(
      `it should print the npm registry username when config has a valid npmAuthToken`,
      makeTemporaryEnv({}, async ({path, run, source}) => {
        await writeFile(`${path}/.yarnrc.yml`, `npmAuthToken: "${AUTH_TOKEN}"\n`);

        let code;
        let stdout;
        let stderr;

        try {
          ({code, stdout, stderr} = await run(`npm`, `whoami`));
        } catch (error) {
          ({code, stdout, stderr} = error);
        }

        expect({code, stdout, stderr}).toMatchSnapshot();
      })
    );

    test(
      `it should print the npm registry username when config has a valid npmAuthIdent`,
      makeTemporaryEnv({}, async ({path, run, source}) => {
        await writeFile(`${path}/.yarnrc.yml`, `npmAuthIdent: "${AUTH_IDENT}"\n`);

        let code;
        let stdout;
        let stderr;

        try {
          ({code, stdout, stderr} = await run(`npm`, `whoami`));
        } catch (error) {
          ({code, stdout, stderr} = error);
        }

        expect({code, stdout, stderr}).toMatchSnapshot();
      })
    );

    test(
      `it should print the npm registry username for a given scope`,
      makeTemporaryEnv({}, async ({path, run, source}) => {
        const url = await startPackageServer();

        await writeFile(`${path}/.yarnrc.yml`, [
          `npmRegistryServer: "${url}"\n`,
          `npmAuthToken: "${AUTH_TOKEN}"\n`,
          `npmScopes:\n`,
          `  testScope:\n`,
          `    npmAuthToken: ${AUTH_TOKEN2}\n`,
          `    npmRegistryServer: "${url}"\n`,
        ].join(``));

        let code;
        let stdout;
        let stderr;

        try {
          ({code, stdout, stderr} = await run(`npm`, `whoami`, `--scope`, `testScope`));
        } catch (error) {
          ({code, stdout, stderr} = error);
        }

        expect({code, stdout, stderr}).toMatchSnapshot();
      })
    );

    test(
      `it should throw an error when no auth config is found`,
      makeTemporaryEnv({}, async ({path, run, source}) => {
        await expect(run(`npm`, `whoami`)).rejects.toThrowError(/No authentication configured/);
      })
    );

    test(
      `it should throw an error when config has an invalid npmAuthToken`,
      makeTemporaryEnv({}, async ({path, run, source}) => {
        await writeFile(`${path}/.yarnrc.yml`, `npmAuthToken: "${INVALID_AUTH_TOKEN}"\n`);
        await expect(run(`npm`, `whoami`)).rejects.toThrowError(/Invalid authentication \(as an unknown user\)/);
      })
    );

    test(
      `it should throw an error when config has an invalid npmAuthIdent`,
      makeTemporaryEnv({}, async ({path, run, source}) => {
        await writeFile(`${path}/.yarnrc.yml`, `npmAuthIdent: "${INVALID_AUTH_IDENT}"\n`);
        await expect(run(`npm`, `whoami`)).rejects.toThrowError(/Invalid authentication \(as an unknown user\)/);
      })
    );

    test(
      `it should thow an error when invalid auth config is found for a scope`,
      makeTemporaryEnv({}, async ({path, run, source}) => {
        const url = await startPackageServer();

        await writeFile(`${path}/.yarnrc.yml`, [
          `npmScopes:\n`,
          `  testScope:\n`,
          `    npmRegistryServer: "${url}"\n`,
          `npmRegistries:\n`,
          `  "${url}":\n`,
          `    npmAuthToken: ${INVALID_AUTH_TOKEN}`,
        ].join(``));

        await expect(run(`npm`, `whoami`, `--scope`, `testScope`)).rejects.toThrowError(/Invalid authentication \(as an unknown user\)/);
        await expect(run(`npm`, `whoami`)).rejects.toThrowError(/Invalid authentication \(as an unknown user\)/);
      })
    );
  });
});
