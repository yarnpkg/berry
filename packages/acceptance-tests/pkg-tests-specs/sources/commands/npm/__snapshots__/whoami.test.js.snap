// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Commands npm whoami it should print the npm registry username for a given scope 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ YN0000: anotherTestUser
➤ YN0000: Done
",
}
`;

exports[`Commands npm whoami it should print the npm registry username when config has a valid npmAuthIdent 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ YN0000: username
➤ YN0000: Done
",
}
`;

exports[`Commands npm whoami it should print the npm registry username when config has a valid npmAuthToken 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ YN0000: testUser
➤ YN0000: Done
",
}
`;
